external/openglcts/modules/common/glcPixelStorageModesTests.cpp

```diff
@@ -0,0 +1,2498 @@
+
+   generateSrcData();
+   createTexture();
+   createShader();
+   dst = renderToSurf();
+
+   pass = verify(dst, &errMask);
+
+   if (pass) {
+      m_testCtx.getLog()
+      << tcu::TestLog::Message << "Image is valid" << tcu::TestLog::EndMessage
+      << tcu::TestLog::ImageSet("ImageVerification", "Image verification")
+      << tcu::TestLog::Image("Result", "Rendered result", dst.getAccess())
+      << tcu::TestLog::EndImageSet;
+   } else {
+      m_testCtx.getLog()
+      << tcu::TestLog::Message << "Image is invalid" << tcu::TestLog::EndMessage
+      << tcu::TestLog::ImageSet("ErrorVerification", "Image verification")
+      << tcu::TestLog::Image("Result", "Rendered result", dst.getAccess())
+      << tcu::TestLog::Image("ErrorMask", "Error mask", errMask.getAccess())
+      << tcu::TestLog::EndImageSet;
+   }
+
+   m_testCtx.setTestResult(pass ? QP_TEST_RESULT_PASS : QP_TEST_RESULT_FAIL,
+                           pass ? "Pass" : "Fail");
+
+   delete[] m_src_data;
+   delete prog;
+
+   return STOP;
+}
+
+void TexImage2DCase::generateSrcData()
+{
+   m_src_data = new glw::GLubyte[m_cuboid_w * m_cuboid_h * m_pixelsize]();
+
+   glw::GLdouble col = 0.0;
+
+   for (int y = 0; y < m_cuboid_h; y++) {
+      for (int x = 0; x < m_cuboid_w; x++) {
+         if (inrange(y, m_subcuboid_y0, m_subcuboid_y0 + m_subcuboid_h) &&
+             inrange(x, m_subcuboid_x0, m_subcuboid_x0 + m_subcuboid_w))
+            col = 1.0;
+         else
+            col = 0.0;
+         int offset = y * m_cuboid_w * m_pixelsize +
+                      x * m_pixelsize;
+         writePixel(m_src_data + offset, col);
+      }
+   }
+}
+
+void TexImage2DCase::createTexture (void)
+{
+   glu::CallLogWrapper gl (m_context.getRenderContext().getFunctions(),
+                           m_testCtx.getLog());
+   deUint32 tex;
+
+   gl.glGenTextures(1, &tex);
+   gl.glBindTexture(GL_TEXTURE_2D, tex);
+   gl.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+   gl.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+
+   gl.glPixelStorei(GL_UNPACK_ROW_LENGTH,  m_cuboid_w);
+   gl.glPixelStorei(GL_UNPACK_SKIP_ROWS,   m_subcuboid_y0);
+   gl.glPixelStorei(GL_UNPACK_SKIP_PIXELS, m_subcuboid_x0);
+
+   gl.glTexImage2D(GL_TEXTURE_2D,
+                  0,
+                  m_internalFormat,
+                  m_subcuboid_w,
+                  m_subcuboid_h,
+                  0,
+                  m_format,
+                  m_type,
+                  m_src_data);
+
+   gl.glPixelStorei(GL_UNPACK_ROW_LENGTH,  0);
+   gl.glPixelStorei(GL_UNPACK_SKIP_PIXELS, 0);
+   gl.glPixelStorei(GL_UNPACK_SKIP_ROWS,   0);
+}
+
+void TexImage2DCase::createShader (void)
+{
+   glu::CallLogWrapper gl (m_context.getRenderContext().getFunctions(),
+                           m_testCtx.getLog());
+   char vs_src[1024];
+   char fs_src[1024];
+
+   char ver[8];
+   glu::ContextType ctx_type = m_context.getRenderContext().getType();
+   sprintf(ver, "%d%s",
```

Comment:
Please do not use sprintf or similar functions. C++ string concatenations should suffice in this case.
You shouldn't deduce the version string from the current context version. The GLSL version must be the min you need e.g. #version 310 es or #version  450. And you should just select the appropriate one depending on the test suite.